test/catch2/ext/test_key_provider.cpp

```diff
@@ -0,0 +1,265 @@
+/*-
+ * Copyright (c) 2014-present MongoDB, Inc.
+ * Copyright (c) 2008-2014 WiredTiger, Inc.
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#include "ext/test/key_provider/key_provider.h"
+
+#include "wrappers/mock_session.h"
+#include "utils.h"
+
+#include "wiredtiger.h"
+#include "wt_internal.h"
+
+#include <catch2/catch.hpp>
+
+#include <functional>
+#include <iostream>
+#include <memory>
+
+/*
+ * kp_fixture
+ *     Test fixture for the key provider extension tests.
+ */
+struct kp_fixture {
+    utils::shared_library lib{KEY_PROVIDER_EXTENSION};
+
+    using extension_init_t = decltype(&wiredtiger_extension_init);
+    extension_init_t extension_init = nullptr;
+
+    static constexpr char conn_config[] = "create,in_memory";
+    WT_CONNECTION *conn = nullptr;
+
+    KEY_PROVIDER *kp = nullptr;
+    using kp_ptr_t = std::unique_ptr<KEY_PROVIDER, std::function<void(KEY_PROVIDER *)>>;
+
+    ~kp_fixture()
+    {
+        reset();
+
+        if (conn != nullptr) {
+            conn->close(conn, nullptr);
+            conn = nullptr;
+        }
+    }
+
+    kp_fixture() : extension_init(lib.get<extension_init_t>("wiredtiger_extension_init"))
+    {
+        int ret = wiredtiger_open(DB_HOME, nullptr, conn_config, &conn);
+        REQUIRE(ret == 0);
+        REQUIRE(conn != nullptr);
+    }
+
+    kp_ptr_t
+    setup(const char *config)
+    {
+        const char *ext_config[] = {config, nullptr};
+
+        int ret = extension_init(conn, reinterpret_cast<WT_CONFIG_ARG *>(ext_config));
+        REQUIRE(ret == 0);
+
+        REQUIRE(kp == nullptr);
+        kp = reinterpret_cast<KEY_PROVIDER *>(wt_conn_impl()->key_provider);
+        REQUIRE(kp != nullptr);
+
+        return kp_ptr_t(kp, [this](KEY_PROVIDER *k) {
+            assert(k == this->kp);
+            this->reset();
+        });
+    }
+
+    void
+    reset()
```

Comment:
np: i'm thinking setup and teardown/destroy would be better word choices? WDYT?

Reply:
I rather rename `setup()` to something like `load()` or `init()`. Both methods can be called arbitrarily, and there is no guarantee that any of them will be called at test start/end.

The `reset()` method is to follow `std::unique_ptr::reset()` semantics. To indicate that internal instance of the extension object was cleared.